feat(Core/Grids): Remove WorldObject separation in grid containers

Takenbacon opened the pull request:
<!-- First of all, THANK YOU for your contribution. -->

Removes the arbitrary separation of world and grid objects in the grid container that no one is able to figure out why it exists or is necessary. This is definitely on the more experimental side and I'm unsure of side effects, needs very proper testing before merge.

May see a very slight performance increase in visitors that previously visited both container types and a small reduction in memory usage.

## Changes Proposed:
<!-- If your pull request promotes complex changes that require a detailed explanation, please describe them in detail specifying what your solution is and what is it meant to address. -->
This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:
<!-- If your fix has a relating issue, link it below -->

## SOURCE:
<!-- If you can, include a source that can strengthen your claim -->
The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:
<!-- Does it build without errors? Did you test in-game? What did you test? On which OS did you test? Describe any other tests performed -->
This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [X] This pull request requires further testing and may have edge cases to be tested.


## How to Test the Changes:
<!-- Describe in a detailed step-by-step order how to test the changes -->

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

- [ ]
- [ ]

<!-- If you intend to contribute repeatedly to our project, it is a good idea to join our discord channel. We set ranks for our contributors and give them access to special resources or knowledge: https://discord.com/invite/GyFvXpk7)
     Do not remove the instructions below about testing, they will help users to test your PR -->
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

sudlud commented:
leaving this here for reference, as far as i can tell the split into grid / world lists originates here:
https://github.com/mangos/mangos-svn/commit/79a135e29ffd7de0e9f79a7f6e8706fe99f1ec92

which was then e.g. used addressing issues with resurrectable corpses:
https://github.com/mangos/mangos-svn/commit/6a04c0a4531ffc72c61c1cb05bf1604687133bf5

and probably to address some pet issues:
https://github.com/mangos/mangos-svn/commit/1a6dd27fca13c73070832b01e7b431502ab8aeee

but all in all i agree with @Takenbacon that the reasons/issues why it got implemented like this are either not valid anymore today or are lost in history

PkllonG commented:
As an aside, is there any way to save PRs in web pages as .patch files?

Takenbacon commented:
> There is a conflict between PR？

Looks to be now after other code was merged. I'll fix shortly.

Takenbacon commented:
> I like this change and want it in TrinityCore too but right now this would be a performance regression for MessageDistDeliverer because it needs to visit creatures container to deal with shared vision stuff - instead of visiting a handful of creatures and pets, it will need to check all creatures (thats what the container switching was for)

You do make an excellent point, I had not considered that. Reworking visibility to remove the use of grid searchers was on my todo list, guess I'll be working on that next then.

Takenbacon commented:
Updated to master and removed draft, should be ready for review/testing now after the merged visibility changes to master.

CI checks needs to be re-triggered, looks like they don't run while the PR was marked as draft.

PkllonG commented:
I feel a noticeable performance gain
